pkg/cri/io/metrics.go

```diff
@@ -0,0 +1,42 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package io
+
+import "github.com/docker/go-metrics"
+
+var (
+	inputEntries  metrics.Counter
+	outputEntries metrics.Counter
+	inputBytes    metrics.Counter
+	outputBytes   metrics.Counter
+	splitEntries  metrics.Counter
+)
+
+func init() {
+	// These CRI metrics record input and output logging volume.
+	ns := metrics.NewNamespace("containerd", "cri", nil)
```

Comment:
Dumb q for someone unfamiliar with prometheus (and more so this docker wrapper around the client), is registering the same subsystem twice valid with no weird side effects (I assume so if this built 😆)? I see we register the same name and subsystem back in [../server](https://github.com/containerd/containerd/blob/3bfa8556cfb92f0ac1f91caef95275185f2fe5e3/pkg/cri/server/metrics.go#L41)

Reply:
That's a good point. I didn't get build error and both /server metrics and logging metrics defined here can be collected successfully. Here is an example

```
# TYPE containerd_cri_container_list_seconds histogram
containerd_cri_container_list_seconds_bucket{le="0.005"} 1453
containerd_cri_container_list_seconds_bucket{le="0.01"} 1454
containerd_cri_container_list_seconds_bucket{le="0.025"} 1454
...
# HELP containerd_cri_input_bytes_total Size of logs received
# TYPE containerd_cri_input_bytes_total counter
containerd_cri_input_bytes_total 3242
# HELP containerd_cri_input_entries_total Number of log entries received
# TYPE containerd_cri_input_entries_total counter
containerd_cri_input_entries_total 29
...
# HELP containerd_cri_sandbox_create_network_seconds time to create the network for a sandbox
# TYPE containerd_cri_sandbox_create_network_seconds histogram
containerd_cri_sandbox_create_network_seconds_bucket{le="0.005"} 0
containerd_cri_sandbox_create_network_seconds_bucket{le="0.01"} 0
```

Looking at implementation of the docker wrapper and Prometheus, the Register requires each description in the description channel to be unique (https://github.com/prometheus/client_golang/blob/170205fb58decfd011f1550d4cfb737230d7ae4f/prometheus/registry.go#L284-L293). It calls Describe() method of a collector to get description. The namespace in docker wrapper is the collector and its description is generated from description of each individual metrics (https://github.com/docker/go-metrics/blob/b619b3592b65de4f087d9f16863a7e6ff905973c/namespace.go#L127-L134). I wonder if that's why I can register the same subsystem twice.

That's being said, if you would suggest having a different name for subsystem (e.g., cri-logging) to be safe and/or distinguish with metrics from /server, I'm good with that too. I am not sure what is the requirement for naming subsystems.

Reply:
If both events collect just fine it seems alright to leave it. The description for the metrics also gives away that it's for logging so that's good also, thanks!